handler/instance-started_test.go

```diff
@@ -1,38 +1,272 @@
 package handler_test
-/*
+
 import (
-	"os"
 	"testing"
 	"context"
+	"time"
+	"net/http"
+	"errors"
+
+	"github.com/ONSdigital/dp-api-clients-go/dataset"
+	"github.com/ONSdigital/dp-api-clients-go/recipe"
+	"github.com/ONSdigital/dp-api-clients-go/cantabular"
 
 	"github.com/ONSdigital/dp-import-cantabular-dataset/config"
 	"github.com/ONSdigital/dp-import-cantabular-dataset/handler"
 	"github.com/ONSdigital/dp-import-cantabular-dataset/event"
+	"github.com/ONSdigital/dp-import-cantabular-dataset/schema"
+
+	"github.com/ONSdigital/dp-kafka/v2/kafkatest"
+	"github.com/ONSdigital/dp-import-cantabular-dataset/handler/mock"
 
 	. "github.com/smartystreets/goconvey/convey"
 )
-var (
-	testEvent = event.InstanceStarted{
-		RecipientName: "World",
-	}
-)
 
-func TestInstanceStartedHandler_Handle(t *testing.T) {
+type testError struct{
+	statusCode int
+}
+
+func (e *testError) Error() string {
+	return "I am a test error"
+}
+
+func (e *testError) StatusCode() int {
+	return e.statusCode
+}
+
+type statusCoder interface{
+	StatusCode() int
+}
+
+func TestInstanceStartedHandler_HandleHappy(t *testing.T) {
+	cfg := config.Config{}
+
+	ctblrClient := cantabularClientHappy()
+	recipeAPIClient := recipeAPIClientHappy()
+	datasetAPIClient := datasetAPIClientHappy()
+	producer := kafkatest.NewMessageProducer(true)
+
+	h := handler.NewInstanceStarted(
+		cfg,
+		&ctblrClient,
+		&recipeAPIClient,
+		&datasetAPIClient,
+		producer,
+	)
+	ctx := context.Background()
+
 	Convey("Given a successful event handler, when Handle is triggered", t, func() {
-		ctx := context.Background()
-		h := &handler.InstanceStarted{}
+		input := make(chan []byte, 2)
+		go func(){
+			for {
+				select{
+					case b := <-producer.Channels().Output:
+						input <- b
+					case <-time.After(time.Second * 1):
+						return
+				}
+			}
+		}()
 
-		err := h.Handle(ctx, &config.Config{OutputFilePath: filePath}, &testEvent)
+		err := h.Handle(ctx, &event.InstanceStarted{
+			RecipeID: "test-recipe-id",
+			InstanceID: "test-instance-id",
+			JobID: "test-job-id",
+			CantabularType: "cantabular-table",
+		})
 		So(err, ShouldBeNil)
+
+		time.Sleep(time.Second * 1)
```

Comment:
I usually try to avoid sleeping in the unit tests. Mainly, to avoid race conditions (we observed some problems in the past, where we tested graceful shutdowns by introducing sleeps and had race conditions).

I would suggest using a wait group, I tested the following and it would work:

```
Convey("Given a successful event handler, when Handle is triggered", t, func(c C) {

		// trigger a go-routine that validates the messages sent to the producer
		wg := &sync.WaitGroup{}
		wg.Add(1)
		go func() {
			defer wg.Done()

			// validate first message
			select {
			case b := <-producer.Channels().Output:
				validateMessage(b, c, event.CategoryDimensionImport{
					JobID:       "test-job-id",
					InstanceID:  "test-instance-id",
					DimensionID: "test-variable",
				})
			case <-time.After(testTimeout):
				t.Fail()
			}

			// validate second message
			select {
			case b := <-producer.Channels().Output:
				validateMessage(b, c, event.CategoryDimensionImport{
					JobID:       "test-job-id",
					InstanceID:  "test-instance-id",
					DimensionID: "test-mapped-variable",
				})
			case <-time.After(testTimeout):
				t.Fail()
			}
		}()

		// execute the Handler under test
		err := h.Handle(ctx, &event.InstanceStarted{
			RecipeID:       "test-recipe-id",
			InstanceID:     "test-instance-id",
			JobID:          "test-job-id",
			CantabularType: "cantabular-table",
		})
		So(err, ShouldBeNil)

		// wait for the go-routine to finish its execution and validation
		wg.Wait()

		err = producer.Close(ctx)
		So(err, ShouldBeNil)
	})
```

Where validateMessage is:
```
// validateMessage unmarshals the provided array of bytes into a CategoryDimensionImport event
// and validates that it resembles the provied expected event
func validateMessage(receivedBytes []byte, c C, expectedEvent event.CategoryDimensionImport) {
	var got event.CategoryDimensionImport
	s := schema.CategoryDimensionImport
	err := s.Unmarshal(receivedBytes, &got)
	c.So(err, ShouldBeNil)
	c.So(got, ShouldResemble, expectedEvent)
}
```
(note that I had to introduce c C parameter in the convey func, to be able to validate things in a different go-routine)

Reply:
Update: 
Sorry, the select wouldn't guarantee order. Please feel free to reintroduce the `got` slice (inside the go-routine, I would suggest, and keeping the wait group)

Reply:
If the selects introduce too much code, maybe you can move it to the validate func.

Reply:
Yeah using a sleep is definitely a bit scrappy, using a wait group like this would be much cleaner